### PR TITLE
Optimize JSON_SET and JSON_REPLACE on `IndexedJsonDocument`

### PR DESCRIPTION
**Pull request by nicktobey**
This PR includes a new implementation of the JSON_SET and JSON_REPLACE functions that leverage the new indexed JSON storage format.

For JSON documents that span multiple chunks, only the affected chunks need to be loaded and modified, allowing operations to scale with the size of the removed value instead of the size of the entire document.

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@nicktobey __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d7cb874</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d7cb874</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Review by fulghum — APPROVED** [review]
Looks good! Just a couple comments about two pieces of duplicated code that I noticed.

File: go/store/prolly/tree/json_indexed_document.go [hunk 1]
```diff
@@ -386,22 +385,130 @@ func (i IndexedJsonDocument) tryRemove(ctx context.Context, path string) (types.
...
 	}
-	return types.JSONDocument{Val: v}.Set(ctx, path, val)
+
+	jsonCursor, found, err := newJsonCursor(ctx, i.m.NodeStore, i.m.Root, keyPath, false)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// The supplied path may be 0-indexing into a scalar, which is the same as referencing the scalar. Remove
+	// the index and try again.
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_indexed_document.go:416]: There's definitely cases that we don't cover here, such as including the [0] index in the middle of a path.

Example: the number in `{ "a": { "b": 1 } }` could be accessed via the path `$[0].a[0].b[0]`.

This isn't the right place for a comprehensive solution, this is just to handle some cases that we currently test for. A comprehensive solution would go somewhere else, but I don't think is worth implementing unless a user actually needs this behavior, given how odd it is.

File: go/store/prolly/tree/json_indexed_document_test.go [hunk 2]
```diff
@@ -259,6 +259,28 @@ func TestIndexedJsonDocument_Extract(t *testing.T) {
 	jsontests.RunJsonTests(t, testCases)
 }
 
+func TestIndexedJsonDocument_Replace(t *testing.T) {
+	ctx := context.Background()
+	ns := NewTestNodeStore()
+	convertToIndexedJsonDocument := func(t *testing.T, s interface{}) interface{} {
+		return newIndexedJsonDocumentFromValue(t, ctx, ns, s)
+	}
```
[the remark this answers is not in the document]
Comment (nicktobey) [reply go/store/prolly/tree/json_indexed_document_test.go:267]: The function is a closure that captures the context and node store from the surrounding block. Pulling it out makes that more complicated and doesn't seem worth it.

**Pull request merged**